src/theory/theory_model_builder.cpp

```diff
@@ -389,124 +388,197 @@ bool TheoryEngineModelBuilder::buildModel(TheoryModel* tm)
+  eq::EqualityEngine* ee = tm->d_equalityEngine;
+  NodeSet assignableCache;
+  std::map<Node, Node>::iterator itm;
+  eq::EqClassesIterator eqcs_i = eq::EqClassesIterator(ee);
+  // should we compute assigner objects?
+  bool computeAssigners = tm->hasAssignmentExclusionSets();
+  // the set of exclusion sets we have processed
+  std::unordered_set<Node, NodeHashFunction> processedExcSet;
   for (; !eqcs_i.isFinished(); ++eqcs_i)
   {
-    // eqc is the equivalence class representative
-    Node eqc = (*eqcs_i);
-    Trace("model-builder") << "Processing EC: " << eqc << endl;
-    Assert(tm->d_equalityEngine->getRepresentative(eqc) == eqc);
-    TypeNode eqct = eqc.getType();
-    Assert(assertedReps.find(eqc) == assertedReps.end());
-    Assert(d_constantReps.find(eqc) == d_constantReps.end());
+    Node eqc = *eqcs_i;
+
+    // Information computed for each equivalence class
+
+    // The assigned represenative and constant representative
+    Node rep, constRep;
+    // A flag set to true if the current equivalence class is assignable (see
+    // assignableEqc).
+    bool assignable = false;
+    // Set to true if the current equivalence class is evaluatable (see
+    // evaluableEqc).
+    bool evaluable = false;
+    // Set to true if a term in the current equivalence class has been given an
+    // assignment exclusion set.
+    bool hasESet CVC4_UNUSED = false;
+    // Set to true if we found that a term in the current equivalence class has
+    // been given an assignment exclusion set, and we have not seen this term
+    // as part of a previous assignment exclusion group. In other words, when
+    // this flag is true we construct a new assigner object with the current
+    // equivalence class as its master.
+    bool foundESet = false;
+    // The assignment exclusion set for the current equivalence class.
+    std::vector<Node> eset;
+    // The group to which this equivalence class belongs when exclusion sets
+    // were assigned (see the argument group of
+    // TheoryModel::getAssignmentExclusionSet).
+    std::vector<Node> esetGroup;
 
     // Loop through terms in this EC
-    Node rep, const_rep;
-    eq::EqClassIterator eqc_i = eq::EqClassIterator(eqc, tm->d_equalityEngine);
+    eq::EqClassIterator eqc_i = eq::EqClassIterator(eqc, ee);
     for (; !eqc_i.isFinished(); ++eqc_i)
     {
       Node n = *eqc_i;
       Trace("model-builder") << "  Processing Term: " << n << endl;
-      // Record as rep if this node was specified as a representative
-      if (tm->d_reps.find(n) != tm->d_reps.end())
+
+      // For each term n in this equivalence class:
+      // (1) Register its assignable subterms,
+      // (2) Compute whether it is a constant or assigned representative,
+      // (3) If we don't have a constant representative, compute information
+      // regarding how we will assign values.
+
+      // (1) Add assignable subterms, which ensures that e.g. models for
```

Comment:
Is this (1) the same as the (1) above? If so, I would try to avoid writing it twice in different words. Maybe move all the text to above, and here just write (1)? Same for the other numbers before.

Reply:
Good point, I'll clean this up.